### PR TITLE
Fetching deposits by range

### PR DESCRIPTION
**Pull request by austinabell**
- Implements #321 

I added some basic unit tests but can add any other edge cases if you guys want but it didn't seem necessary. Didn't seem necessary to test the result of `deposit.test.ts` I added, but can add that in (just seems irrelevant since the `.resolves()` was literally setting the return value as I found out the hard way lol)

As part of this issue would you like me to connect this to any functionality or possibly a new CLI command?

### CODE REVIEW

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Sorry I was busy earlier to answer your question in issue. A bit of context, deposits are comming in regulary and after they are included in block they will get deleted. So imagine scenario where you have in db deposits with index 5,6,7,8. When creating new block, you want blocks from state.eth1DepositIndex(6) till min(MAX_DEPOSITS, votedEth1Data.depositCount(7)). So if I pass 5 and 6 as params to your function I won't get anything. Since deposit index is part of db key, you could do

```
gt: encodeKey(this.bucket, lowerIndex),
lt: encodeKey(this.bucket, uperIndex),
```
and it should return deposits with that index. Might be good idea to write e2e test to test that theory.

Hope I made clearer.

**Comment by mpetrunic** [comment]
@austinabell also, that method should be used in src/chain/factory/block/deposits.ts (you have TODO)

File: packages/lodestar/src/db/api/beacon/repository.ts [hunk 1]
```diff
@@ -64,15 +64,10 @@ export abstract class BulkRepository<T> extends Repository<T> {
   }
 
   public async getAllBetween(lowerLimit: number, upperLimit: number): Promise<T[]> {
-    let data = await this.db.search({
-      gt: encodeKey(this.bucket, Buffer.alloc(0)),
-      lt: encodeKey(this.bucket + 1, Buffer.alloc(0)),
+    const data = await this.db.search({
+      gt: encodeKey(this.bucket, Buffer.alloc(lowerLimit)),
```
Comment (mpetrunic) [inline packages/lodestar/src/db/api/beacon/repository.ts]: remove buffer alloc from gt and lt, encodeKey accepts number @austinabell 

Just so you don't debug too much why this doesnt work :smile:

Comment (austinabell) [reply packages/lodestar/src/db/api/beacon/repository.ts]: thanks! Just kept as Buffer since that's what it was previously and I saw this could accept a Buffer

**Review by GregTheGreek — DISMISSED** [review]

**Comment by austinabell** [comment]
Please don't merge in yet, going to finish tests for this PR when I get in

File: packages/lodestar/src/db/api/beacon/repository.ts [hunk 2]
```diff
@@ -65,7 +65,7 @@ export abstract class BulkRepository<T> extends Repository<T> {
 
   public async getAllBetween(lowerLimit: number, upperLimit: number): Promise<T[]> {
     const data = await this.db.search({
-      gt: encodeKey(this.bucket, lowerLimit),
+      gt: encodeKey(this.bucket, lowerLimit || Buffer.alloc(0)),
```
Comment (mpetrunic) [inline packages/lodestar/src/db/api/beacon/repository.ts:70]: should be just 0 not buffer alloc

**Review by mpetrunic — CHANGES_REQUESTED** [review]
I wouldn't merge it before e2e test to make sure this idea works

File: packages/lodestar/src/db/api/beacon/repository.ts (hunk 2 above)
Comment (austinabell) [reply packages/lodestar/src/db/api/beacon/repository.ts:70]: With that it didn't return the 0th index element for some reason. I needed to add this to have it be included. There wasn't a good way to set up a unit test but if you make that change and check the values returned in the e2e test, the 0th index was never included.

**Review by wemeetagain — APPROVED** [review]

File: packages/lodestar/test/e2e/opPool/modules/deposit.test.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  });
+
+  it("should fetch deposit from range - upper and lower limit",async function () {
+    await Promise.all([
+      db.deposit.set(0, generateDeposit()),
+      db.deposit.set(2, generateDeposit()),
+      db.deposit.set(4, generateDeposit()),
+      db.deposit.set(6, generateDeposit())
+    ]);
+    const result = await opPool.deposits.getAllBetween(1, 7);
```
Comment (austinabell) [inline packages/lodestar/test/e2e/opPool/modules/deposit.test.ts:57]: Just want to confirm that the range should be bounds not inclusive, because changing this test to between 2-7, only 4 and 6 are returned. This was the issue I was running into. This will be fine when doing a range starting at 0 because `lowerLimit || Buffer.alloc(0)` will choose the Buffer.alloc(0), but for any other value, the first index will not be included.

By the way sorry I wasn't able to come back to this before you got to it

Comment (mpetrunic) [reply packages/lodestar/test/e2e/opPool/modules/deposit.test.ts:57]: It's fine, I left comment about bounds on method :)

Comment (mpetrunic) [reply packages/lodestar/test/e2e/opPool/modules/deposit.test.ts:57]: The single limit tests demonstrate that it won't include boundaries

Comment (austinabell) [reply packages/lodestar/test/e2e/opPool/modules/deposit.test.ts:57]: okay awesome! just wanted to make sure I wasn't adding technical debt :)

**Review by mpetrunic — APPROVED** [review]
